providers/common/exchange/dh_exch.c

```diff
@@ -42,22 +42,26 @@ static int dh_init(void *vpdhctx, void *vdh)
 {
     PROV_DH_CTX *pdhctx = (PROV_DH_CTX *)vpdhctx;
 
+    if (pdhctx == NULL)
+        return 0;
     DH_free(pdhctx->dh);
     pdhctx->dh = vdh;
-    DH_up_ref(pdhctx->dh);
-
-    return pdhctx->dh != NULL;
+    if (vdh == NULL)
+        return 0;
+    return DH_up_ref(pdhctx->dh);
 }
 
 static int dh_set_peer(void *vpdhctx, void *vdh)
 {
     PROV_DH_CTX *pdhctx = (PROV_DH_CTX *)vpdhctx;
 
+    if (pdhctx == NULL)
+        return 0;
     DH_free(pdhctx->dhpeer);
     pdhctx->dhpeer = vdh;
-    DH_up_ref(pdhctx->dhpeer);
-
-    return pdhctx->dhpeer != NULL;
+    if (vdh == NULL)
+        return 0;
+    return DH_up_ref(pdhctx->dhpeer);
```

Comment:
here as well.  If DH_up_ref fails the pdhctx is left in an undefined state.

Reply:
Likewise.